### PR TITLE
Adding FP32 gating + noisy gating + top2 gating

### PR DESCRIPTION
**Pull request by ykim362**
* Add fp32 gating
* Add noisy selection of gate
* Add top-2 gating back

Acknowledgment: This PR is the result of peer coding sessions with Young and Ammar.

Co-authored-by: Young Jin Kim <youki@microsoft.com>
Co-authored-by: soartist1 <soartist1@gmail.com>

### CODE REVIEW

File: deepspeed/moe/sharded_moe.py [hunk 1]
```diff
@@ -112,14 +192,27 @@ class TopKGate(torch.nn.Module):
...
         super().__init__()
-        self.wg = torch.nn.Linear(model_dim, num_experts, bias=False)
+
+        # Only top-1 and top-2 are supported at the moment.
+        if k != 1 and k != 2:
+            raise ValueError('Only top-1 and top-2 gatings are supported.')
+        self.wg = torch.nn.Linear(model_dim, num_experts, bias=False).float()
+        self.k = k
         self.capacity_factor = capacity_factor
+        self.noisy_gate = noisy_gate
```
Comment (awan-10) [inline deepspeed/moe/sharded_moe.py:205]: TODO: Add a check to make sure noisy_gate is used only for training

**Comment by TopcoderY** [comment]
I notice that in top2gating,router weight normalization is deployed. Is there any reference to this operation?
